src/PIL/ImageDraw.py

```diff
@@ -350,34 +351,43 @@ def floodfill(image, xy, value, border=None, thresh=0):
         pixel[x, y] = value
     except (ValueError, IndexError):
         return  # seed point outside image
-    edge = [(x, y)]
+    edge = {(x, y)}
+    full_edge = set()  # use a set to record each unique pixel processed
     if border is None:
         while edge:
-            newedge = []
-            for (x, y) in edge:
+            new_edge = set()
+            for (x, y) in edge:  # 4 adjacent method
                 for (s, t) in ((x+1, y), (x-1, y), (x, y+1), (x, y-1)):
+                    if (s, t) in full_edge:
+                        continue  # if already processed, skip
                     try:
                         p = pixel[s, t]
                     except IndexError:
                         pass
                     else:
                         if _color_diff(p, background) <= thresh:
                             pixel[s, t] = value
-                            newedge.append((s, t))
-            edge = newedge
+                            new_edge.add((s, t))
+                            full_edge.add((s, t))
+            full_edge = edge  # do not record useless pixels to reduce memory consumption
+            edge = new_edge
     else:
         while edge:
-            newedge = []
+            new_edge = set()
             for (x, y) in edge:
                 for (s, t) in ((x+1, y), (x-1, y), (x, y+1), (x, y-1)):
+                    if (s, t) in full_edge:
+                        continue
                     try:
                         p = pixel[s, t]
                     except IndexError:
                         pass
                     else:
                         if p != value and p != border:
                             pixel[s, t] = value
-                            newedge.append((s, t))
+                            new_edge.add((s, t))
+                            full_edge.add((s, t))
+            full_edge = edge
             edge = newedge
```

Comment:
Thanks for the PR! Before we can check the performance improvements, the PR should be passing for both CIs: Travis CI for Linux and AppVeyor for Windows. Right now they are both failing because this `newedge` is not defined. Rename it to `new_edge` as well?

```console
$ flake8 src/PIL/ImageDraw.py
src/PIL/ImageDraw.py:372:80: E501 line too long (89 > 79 characters)
src/PIL/ImageDraw.py:391:20: F821 undefined name 'newedge'
```

(I wonder why your tests passed? Was it truly testing this modified code or another version?)

Reply:
sorry... a typo was found in the committed file 😓  I should just paste the tested version into it...